### PR TITLE
Add bulk query support, improve tests and fix upsert issue

### PR DESCRIPTION
**Pull request by ThishaniLucas**

### CODE REVIEW

File: src/sfdc/salesforce_bulk_endpoint.bal [hunk 1]
```diff
@@ -174,88 +186,77 @@ public type BulkJob client object {
...
-                    xml payload = check convertToXml(content);
-                    req.setXmlPayload(<@untained>  payload);
-                    var response = self.httpClient->post(path, req);
-                    xml|ConnectorError batchResponse = checkXmlPayloadAndSetErrors(response);
-                    if (batchResponse is xml){
-                        BatchInfo binfo = check createBatchRecordFromXml(batchResponse);
-                        return binfo;
+                    if (QUERY == self.operation) {
+                        string payload = check convertToString(content);
+                        req.setTextPayload(<@untained>  payload);
```
Comment (Bhashinee) [inline src/sfdc/salesforce_bulk_endpoint.bal:222]:
```suggestion
                        req.setTextPayload(<@untainted>  payload);
```
Check all the other places as well.

File: src/sfdc/salesforceBulkUtils/sf_bulk_utils.bal [hunk 2]
```diff
@@ -328,3 +363,80 @@ function convertToXml(io:ReadableByteChannel rbc) returns @tainted xml|Connector
         }
     }
 }
+
+function getJsonQueryResult(json resultlist, string path, http:Client httpClient) returns @tainted json|ConnectorError {
+    json[] finalResults = [];
+    http:Request req = new;
+    //result list is always a json[]
+    if(resultlist is json[]) {
```
Comment (Bhashinee) [inline src/sfdc/salesforceBulkUtils/sf_bulk_utils.bal:371]:
```suggestion
    if (resultlist is json[]) {
```

File: src/sfdc/salesforceBulkUtils/sf_bulk_utils.bal [hunk 3]
```diff
@@ -328,3 +363,80 @@ function convertToXml(io:ReadableByteChannel rbc) returns @tainted xml|Connector
...
+        foreach var item in resultlist {
+            string resultId = item.toString();
+            var response = httpClient->get(path + "/"+ resultId, req);
+            json result = check checkJsonPayloadAndSetErrors(response);
+            //result is always a json[]
+            if(result is json[]) {
```
Comment (Bhashinee) [inline src/sfdc/salesforceBulkUtils/sf_bulk_utils.bal:377]:
```suggestion
            if (result is json[]) {
```

File: src/sfdc/salesforce_bulk_endpoint.bal [hunk 4]
```diff
@@ -174,88 +186,77 @@ public type BulkJob client object {
...
-                    json payload = check convertToJson(content);
-                    req.setJsonPayload(<@untained>  payload);
-                    var response = self.httpClient->post(path, req);
-                    json|ConnectorError batchResponse = checkJsonPayloadAndSetErrors(response);
-                    if (batchResponse is json){
-                        BatchInfo binfo = check BatchInfo.constructFrom(batchResponse);
-                        return binfo;
+                    if (QUERY == self.operation) {
+                        string payload = check convertToString(content);
+                        req.setTextPayload(<@untained>  payload);
```
Comment (Bhashinee) [inline src/sfdc/salesforce_bulk_endpoint.bal:196]:
```suggestion
                        req.setTextPayload(<@untainted>  payload);
```

Check all the other places as well.
https://github.com/ballerina-platform/module-ballerinax-sfdc/pull/102/files#diff-5b542c5854666692d4661bff5f59216fR225
https://github.com/ballerina-platform/module-ballerinax-sfdc/pull/102/files#diff-eb02955ead80dade0dfac32d254213cdR18
https://github.com/ballerina-platform/module-ballerinax-sfdc/pull/102/files#diff-ec972bdf56dc32665948f3f1b261af47R98
https://github.com/ballerina-platform/module-ballerinax-sfdc/pull/102/files#diff-3a4b2095243c560558aa955df684d2c5R23
https://github.com/ballerina-platform/module-ballerinax-sfdc/pull/102/files#diff-9276837cd9f633225aedfc2e5b933789R120
https://github.com/ballerina-platform/module-ballerinax-sfdc/pull/102/files#diff-0521f2f1fa79403950e013198aea0199R116
https://github.com/ballerina-platform/module-ballerinax-sfdc/pull/102/files#diff-5b542c5854666692d4661bff5f59216fR196
https://github.com/ballerina-platform/module-ballerinax-sfdc/pull/102/files#diff-5b542c5854666692d4661bff5f59216fR199

File: src/sfdc/salesforceBulkUtils/sf_bulk_utils.bal [hunk 5]
```diff
@@ -328,3 +363,80 @@ function convertToXml(io:ReadableByteChannel rbc) returns @tainted xml|Connector
...
+                finalResults = mergeCsv(finalResults, result);
+            }
+        } 
+        i = i + 1;      
+    } 
+    return finalResults;  
+}
+
+function mergeJson(json[] list1, json[] list2) returns json[] {
+    foreach var item in list2{
```
Comment (daneshk) [inline src/sfdc/salesforceBulkUtils/sf_bulk_utils.bal:421]:
```suggestion
    foreach var item in list2 {
```

File: src/sfdc/tests/bulkTests/csv/delete.bal [hunk 6]
```diff
@@ -0,0 +1,81 @@
+import ballerina/log;
```
Comment (daneshk) [inline src/sfdc/tests/bulkTests/csv/delete.bal:1]: Shall we add the license header?

File: src/sfdc/tests/bulkTests/xml/delete.bal [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        if (closedJob is JobInfo) {
+            test:assertTrue(closedJob.state == "Closed", msg = "Closing job failed.");
+        } else {
+            test:assertFail(msg = closedJob.detail()?.message.toString());
+        }
+
+    } else {
+        test:assertFail(msg = deleteJob.detail()?.message.toString());
+    }
+}
```
Comment (daneshk) [inline src/sfdc/tests/bulkTests/xml/delete.bal:83]:
```suggestion
}

```

File: src/sfdc/tests/bulkTests/xml/insert.bal [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+        if (closedJob is JobInfo) {
+            test:assertTrue(closedJob.state == "Closed", msg = "Closing job failed.");
+        } else {
+            test:assertFail(msg = closedJob.detail()?.message.toString());
+        }
+
+    } else {
+        test:assertFail(msg = insertJob.detail()?.message.toString());
+    }
+}
```
Comment (daneshk) [inline src/sfdc/tests/bulkTests/xml/insert.bal:187]:
```suggestion
}

```

File: src/sfdc/tests/bulkTests/xml/query.bal [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        if (closedJob is JobInfo) {
+            test:assertTrue(closedJob.state == "Closed", msg = "Closing job failed.");
+        } else {
+            test:assertFail(msg = closedJob.detail()?.message.toString());
+        }
+
+    } else {
+        test:assertFail(msg = queryJob.detail()?.message.toString());
+    }
+}
```
Comment (daneshk) [inline src/sfdc/tests/bulkTests/xml/query.bal:83]:
```suggestion
}

```

File: src/sfdc/tests/bulkTests/xml/update.bal [hunk 10]
```diff
@@ -1,109 +1,107 @@
...
             test:assertFail(msg = closedJob.detail()?.message.toString());
         }
-        
+
     } else {
-        test:assertFail(msg = insertJob.detail()?.message.toString());
+        test:assertFail(msg = updateJob.detail()?.message.toString());
     }
-}
+}
```
Comment (daneshk) [inline src/sfdc/tests/bulkTests/xml/update.bal:107]:
```suggestion
}

```

File: src/sfdc/tests/bulkTests/xml/upsert.bal [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+        if (closedJob is JobInfo) {
+            test:assertTrue(closedJob.state == "Closed", msg = "Closing job failed.");
+        } else {
+            test:assertFail(msg = closedJob.detail()?.message.toString());
+        }
+
+    } else {
+        test:assertFail(msg = upsertJob.detail()?.message.toString());
+    }
+}
```
Comment (daneshk) [inline src/sfdc/tests/bulkTests/xml/upsert.bal:102]:
```suggestion
}

```

**Pull request merged**